### ISSUE TITLE
Wallaby causes indexing of every file in IntelliJ excluded dirs (such as node_modules)

### ISSUE THREAD

**Issue by vjpr** [wallabyjs/public#605]
### Issue description or question

When running `fs_usage | grep idea` I am seeing lots of `lstat` calls to every file in my symlinked node_modules, including their node_modules. This is thousands of files and completely hangs IntelliJ sometimes. These `lstat` calls continue even after I stop Wallaby.

@ArtemGovorov mentioned here that it might be an issue: https://github.com/wallabyjs/public/issues/575

I feel like this use to work without issue, but then I can't remember if that was before I started linking things.

I've removed the links and the issue goes away. But npm linking is integral to how I work. I think it should be a simple fix in the IntelliJ plugin...

Also, in IntelliJ, my `node_modules` dirs are marked as excluded. I didn't do this before I don't think. Perhaps they were in the index before and Wallaby could quickly scan them, but now that they are not, they are being indexed?

Also, in IntelliJ I am using one project containing multiple folders of node.js modules.

```
project
  - foo
  - bar
    - node_modules
      - foo (symlink)
```

Some insight into how the IntelliJ plugin would be super helpful.

Also, I have tried adding variations of ignore patterns to the wallaby config and it did not help:

```
{pattern: '/Users/Vaughan/dev/gulp4/**', ignore: true},
{pattern: 'node_modules/gulp4/**', ignore: true},
```

Just tested in Atom, does not seem to be a problem. Was very quick to load tests.

---

### Code editor or IDE name and version

WebStorm v11
-> IntelliJ IDEA v15
Visual Studio 2015
Visual Studio 2013
Atom v1.x
Visual Studio Code v0.10.x
Sublime Text v3

### OS name and version

Windows
-> OSX
Linux

**Comment by ArtemGovorov** [comment]
I'm not sure I understand your setup to be able to reproduce the issue. Could you please share a project I may use to reproduce the issue with?

**Comment by vjpr** [comment]
Clone this repo https://github.com/wallabyjs/wallaby-babel-es7-sample

- npm install
- Open it in IntelliJ.
- Exclude `node_modules` immediately.
- Close IntelliJ (don't allow it to finish indexing)
- Open project in IntelliJ again.
- Run `fs_usage | grep idea`
- Observe `lstat` calls for every node_module

```
10:27:31.825601  lstat                                  babel-preset-stage-0/node_modules/babel-preset-stage-1/node_modules/babel-plugin-transform-class-constructor-call/node_modules/babel-runtime/core-js/array/push.js    0.000020   idea.2172303
```

IntelliJ should not need to touch these files as node_modules is excluded and I assume the compilation is done outside of IntelliJ in a Node process.

Excluding `node_modules` in IntelliJ/WebStorm is very common practice for large projects.

**Comment by ArtemGovorov** [comment]
After some investigation, it doesn't look like it is wallaby.js specific issue.

In the repo you have mentioned https://github.com/wallabyjs/wallaby-babel-es7-sample:
- create a simple `gulpfile.js`,
- paste the following content:
```javascript
var gulp = require('gulp'), watch = require('gulp-watch');
gulp.task('default', function () {
  return gulp.src('src/**/*.js').pipe(watch('src/**/*.js')).pipe(gulp.dest('build'));
});
```
- run `fs_usage | grep idea`,
- run the default task using IntelliJ gulp runner.

After doing these steps, you may observe the same increased amount of those fs calls. So just like in [this case](https://github.com/wallabyjs/public/issues/567#issuecomment-217116798), it seems to be the way IntelliJ platform works.

**Comment by ArtemGovorov** [comment]
Actually, it doesn't even have to be `gulp-watch`, the gulpfile like this is enough to reproduce the issue:
```javascript
require('gulp').task('default', function (cb) { setTimeout(cb, 10000); });
```
I'm going to report it to IntelliJ now that we know it's the platform issue and will paste the ticket number here as soon as I report it.

**Comment by vjpr** [comment]
> After doing these steps, you may observe the same increased amount of those fs calls.

So you mean that when running any runner, it will attempt to `lstat` every file including excluded dirs?

...I am testing your gulpfile code now.

**Comment by ArtemGovorov** [comment]
> when running any runner

Not sure about any runner and every file including excluded dirs, but gulp runner definitely produces the same result.

**Comment by vjpr** [comment]
> same result

Just to confirm...the same result being that every file in `node_modules` is `lstat`ed when running wallaby?

I have updated 2nd post quite a bit.

**Comment by vjpr** [comment]
So I checked the gulp task. It does not perform `lstat` on excluded dirs.

To reproduce the bug, `npm install` a module with lots of dependencies in your existing repo. (Ensure `node_modules` is excluded).

Then run wallaby again, and it will say "Scanning source files" in the status bar. While this is happening it is `lstat`ing all the files in the newly added module.

E.g.

```
12:17:54.097508  lstat                                  ple/node_modules/babelator/node_modules/yeoman-generator/node_modules/yeoman-environment/node_modules/grouped-queue/node_modules/lodash/internal/assignDefaults.js    0.000016   idea.2310538
12:17:54.097590  lstat                                  ple/node_modules/babelator/node_modules/yeoman-generator/node_modules/yeoman-environment/node_modules/grouped-queue/node_modules/lodash/internal/createExtremum.js    0.000015   idea.2310538
12:17:54.097651  lstat                                  es7-sample/node_modules/babelator/node_modules/yeoman-generator/node_modules/yeoman-environment/node_modules/grouped-queue/node_modules/lodash/internal/getData.js    0.000014   idea.2310538
12:17:54.097711  lstat                                  mple/node_modules/babelator/node_modules/yeoman-generator/node_modules/yeoman-environment/node_modules/grouped-queue/node_modules/lodash/internal/createForEach.js    0.000013   idea.2310538
12:17:54.097771  lstat                                  mple/node_modules/babelator/node_modules/yeoman-generator/node_modules/yeoman-environment/node_modules/grouped-queue/node_modules/lodash/internal/createWrapper.js    0.000014   idea.2310538
12:17:54.097831  lstat                                  sample/node_modules/babelator/node_modules/yeoman-generator/node_modules/yeoman-environment/node_modules/grouped-queue/node_modules/lodash/internal/baseIsMatch.js    0.000013   idea.2310538
12:17:54.097891  lstat                                  ample/node_modules/babelator/node_modules/yeoman-generator/node_modules/yeoman-environment/node_modules/grouped-queue/node_modules/lodash/internal/bindCallback.js    0.000013   idea.2310538
12:17:54.097951  lstat                                  -es7-sample/node_modules/babelator/node_modules/yeoman-generator/node_modules/yeoman-environment/node_modules/grouped-queue/node_modules/lodash/internal/mapSet.js    0.000013   idea.2310538
12:17:54.098022  lstat                                  mple/node_modules/babelator/node_modules/yeoman-generator/node_modules/yeoman-environment/node_modules/grouped-queue/node_modules/lodash/internal/baseEachRight.js    0.000014   idea.2310538
12:17:54.098083  lstat                                  -sample/node_modules/babelator/node_modules/yeoman-generator/node_modules/yeoman-environment/node_modules/grouped-queue/node_modules/lodash/internal/baseSortBy.js    0.000014   idea.2310538
12:17:54.098143  lstat                                  ample/node_modules/babelator/node_modules/yeoman-generator/node_modules/yeoman-environment/node_modules/grouped-queue/node_modules/lodash/internal/cacheIndexOf.js    0.000013   idea.2310538
12:17:54.098204  lstat                                  /node_modules/babelator/node_modules/yeoman-generator/node_modules/yeoman-environment/node_modules/grouped-queue/node_modules/lodash/internal/createBindWrapper.js    0.000014   idea.2310538
12:17:54.098265  lstat                                  -es7-sample/node_modules/babelator/node_modules/yeoman-generator/node_modules/yeoman-environment/node_modules/grouped-queue/node_modules/lodash/internal/mapGet.js    0.000013   idea.2310538
12:17:54.098325  lstat                                  7-sample/node_modules/babelator/node_modules/yeoman-generator/node_modules/yeoman-environment/node_modules/grouped-queue/node_modules/lodash/internal/cachePush.js    0.000014   idea.2310538
12:17:54.098395  lstat                                  -sample/node_modules/babelator/node_modules/yeoman-generator/node_modules/yeoman-environment/node_modules/grouped-queue/node_modules/lodash/internal/baseAssign.js    0.000014
```

After it runs once, the files seem to be cached/indexed and its fine. To reproduce again, you must install another module.

Could you check the Wallaby IntelliJ plugin to see if there is a VirtualFileSystem API call to read all files in project on start?

**Comment by ArtemGovorov** [comment]
> Just to confirm...the same result being that every file in node_modules is lstated when running wallaby?
> I have updated 2nd post quite a bit.

It looks like node modules `lstat`-ed only when running wallaby. I was looking into the `fstat64` calls that are of the same amount when wallaby/gulp is running. Sorry for misreading. I'll investigate and let you know the results.

**Comment by vjpr** [comment]
@ArtemGovorov Just added another message about how to consistently reproduce it.

**Comment by ArtemGovorov** [comment]
Thanks, I see what the issue is now. We do indeed scan the node modules. When they are not excluded, IDE does it before us, so we hit the cache and everything is fine. When they are excluded, then wallaby scan hit the performance (the same way as the IDE scan if the folder is not excluded). 

Taking into account the IDE excluded roots is not an option, because one may be referencing some specific files/modules in node modules from the wallaby config `files` list. What we should do is to make the initial scan smarter by only determining the list of roots to scan from the `files` list. I'll look into this now.

**Comment by vjpr** [comment]
Thanks! I'm happy to test a pre-release version if you need.

**Comment by ArtemGovorov** [comment]
Please try this build [wallaby-intellij-v1.0.62.zip](https://github.com/wallabyjs/public/files/276890/wallaby-intellij-v1.0.62.zip) (`Settings - Plugins - Install plugin from disk`) and let me know how it goes. If it breaks things from you, just get and install the latest version of the plugin from our website the same way.

**Comment by vjpr** [comment]
Works! 😌

Feels so quick now. How does it work?

**Comment by ArtemGovorov** [comment]
It just doesn't scan stuff it doesn't need to based on the analysis of `files` list. However, I found a couple of bugs I need to fix before I release it. 
Did you also check the scenario of installing/linking a module?

**Comment by vjpr** [comment]
> Did you also check the scenario of installing/linking a module?

Tested the same test project I used before including linking and a couple of my larger ones. Its super quick.

**Comment by ArtemGovorov** [comment]
Cool, thanks! You may use this version for now. I'll clean up the bugs that I found and will close the issue when I publish it, you'll then just receive the notification about the newer version and will be able to install it as always right from the IDE.

**Comment by vjpr** [comment]
One potential edge case may be how it handles symlinks when explicitly specified, and nested node_modules.

When developing with npm linked modules I would like to have the option of testing them too.

So this would require not including the nested `node_modules`. Would these patterns be possible, or something to satisfy this use case?

```
files: [
  'node_modules/a-linked-module',
  {pattern: 'node_modules/a-linked-module/node_modules', ignore: true},
]
```

```
files: [
  'node_modules/a-linked-module/**/!(node_modules)',
]
```

**Comment by ArtemGovorov** [comment]
@vjpr These scenarios should not be affected by the changes I've made.
Could you please try this version [wallaby-intellij-v1.0.62.zip](https://github.com/wallabyjs/public/files/277154/wallaby-intellij-v1.0.62.zip)? It does a little bit more `lstat` calls then the previous one as a result of a couple of bug fixes, but should not affect the performance much.

**Comment by vjpr** [comment]
Works well so far.

**Comment by ArtemGovorov** [comment]
Thanks for your help! Published the new plugin version 1.0.63.
